File: lib/languages/nim/function-arguments.nim
# Basic procedure arguments
proc add(a, b: int): int =
  a + b

# Multiple parameters of the same type
proc multiply(a, b: int): int =
  a * b

# Named arguments
proc createPerson(name: string, age: int): tuple[name: string, age: int] =
  (name: name, age: age)

# Default parameters
proc greet(name: string, greeting = "Hello"): string =
  greeting & ", " & name & "!"

# Var parameters (pass by reference)
proc increment(x: var int) =
  x += 1

# Varargs (variable number of arguments)
proc sum(numbers: varargs[int]): int =
  result = 0
  for n in numbers:
    result += n

# Open arrays (array of unspecified length)
proc average(numbers: openArray[int]): float =
  var total = 0
  for n in numbers:
    total += n
  result = total / numbers.len

# Procedure parameters
proc apply(f: proc(x: int): int, x: int): int =
  f(x)

proc double(x: int): int =
  x * 2

# Generic parameters
proc first[T](arr: openArray[T]): T =
  arr[0]

# Parameter unpacking with tuples
proc processPoint(point: tuple[x, y: int]): int =
  point.x + point.y

# Parameter unpacking with objects
type
  Person = object
    name: string
    age: int

proc describe(p: Person): string =
  p.name & " is " & $p.age & " years old"

# Static parameters (compile-time)
proc factorial(n: static int): int =
  when n <= 1:
    1
  else:
    n * factorial(n-1)

when isMainModule:
  echo "add(5, 3): ", add(5, 3)
  echo "multiply(4, 7): ", multiply(4, 7)
  
  let person = createPerson(name = "Alice", age = 30)
  echo "Person: ", person
  
  echo "greet(\"Bob\"): ", greet("Bob")
  echo "greet(\"Charlie\", \"Hi\"): ", greet("Charlie", "Hi")
  
  var x = 10
  increment(x)
  echo "After increment: ", x
  
  echo "sum(1, 2, 3, 4, 5): ", sum(1, 2, 3, 4, 5)
  
  let numbers = @[1, 2, 3, 4, 5]
  echo "average(numbers): ", average(numbers)
  
  echo "apply(double, 5): ", apply(double, 5)
  
  let mixed = @[10, 20, 30, 40]
  echo "first(mixed): ", first(mixed)
  
  let point = (x: 10, y: 20)
  echo "processPoint(point): ", processPoint(point)
  
  let p = Person(name: "Dave", age: 35)
  echo "describe(p): ", describe(p)
  
  const fact5 = factorial(5)  # Computed at compile time
  echo "factorial(5): ", fact5

File: lib/languages/go/function-arguments.go
package main

import "fmt"

// Basic function with parameters
func greet(name string, age int) {
    fmt.Printf("Hello, %s! You are %d years old.\n", name, age)
}

// Function with same-type parameters
func add(a, b int) int {
    return a + b
}

// Function with variadic parameter (variable number of arguments)
func sum(numbers ...int) int {
    total := 0
    for _, num := range numbers {
        total += num
    }
    return total
}

// Function with pointer parameter
func increment(value *int) {
    *value++
}

// Function with slice parameter (reference type)
func modifySlice(slice []int) {
    if len(slice) > 0 {
        slice[0] = 999 // This will affect the original slice
    }
}

// Function with map parameter (reference type)
func addToMap(m map[string]int, key string, value int) {
    m[key] = value // This will affect the original map
}

// Function with struct parameter
type Person struct {
    Name string
    Age  int
}

// Pass by value (copy)
func celebrateBirthday(p Person) Person {
    p.Age++
    return p // Return modified copy
}

// Pass by pointer (reference)
func celebrateBirthdayPtr(p *Person) {
    p.Age++ // Modifies the original
}

// Function with function parameter (higher-order function)
func applyAndPrint(value int, fn func(int) int) {
    result := fn(value)
    fmt.Println("Result:", result)
}

func main() {
    // Basic parameter passing
    greet("Gopher", 10)

    // Passing arguments by position
    sum := add(5, 3)
    fmt.Println("Sum:", sum)

    // Passing to variadic function
    total := sum(1, 2, 3, 4, 5)
    fmt.Println("Total:", total)

    // Passing a slice to a variadic function
    numbers := []int{1, 2, 3, 4, 5}
    total = sum(numbers...) // Spread operator
    fmt.Println("Total from slice:", total)

    // Passing a pointer
    value := 10
    increment(&value) // Pass address
    fmt.Println("Incremented value:", value)

    // Passing a slice (reference type)
    slice := []int{1, 2, 3}
    fmt.Println("Before:", slice)
    modifySlice(slice)
    fmt.Println("After:", slice) // [999 2 3]

    // Passing a map (reference type)
    scores := map[string]int{"Alice": 95}
    fmt.Println("Before:", scores)
    addToMap(scores, "Bob", 87)
    fmt.Println("After:", scores) // map[Alice:95 Bob:87]

    // Passing a struct by value
    person := Person{Name: "Alice", Age: 30}
    fmt.Println("Before:", person)
    updatedPerson := celebrateBirthday(person)
    fmt.Println("Original after call:", person)        // Age is still 30
    fmt.Println("Returned value:", updatedPerson)      // Age is 31

    // Passing a struct by pointer
    fmt.Println("Before:", person)
    celebrateBirthdayPtr(&person)
    fmt.Println("After:", person) // Age is now 31

    // Passing a function as argument
    applyAndPrint(5, func(x int) int {
        return x * x
    })
}

